tools/compare.go

```diff
@@ -15,125 +15,291 @@
-	ctx := cmd.Context()
+func getBundleFloor(num uint64) uint64 {
+	return uint64(math.Round(float64(num/100.0))) * 100
+}
 
-	storeADef := args[0]
-	storeBDef := args[1]
+func getBundleCeiling(num uint64) uint64 {
+	return (uint64(math.Round(float64(num/100.0))) * 100) + 100
+}
 
-	blockNum, err := strconv.ParseUint(args[2], 10, 64)
+func unifiedDiff(cnt1, cnt2 []byte) (string, error) {
+	file1 := "/tmp/block-difference-expected-bundle"
+	file2 := "/tmp/block-difference-received-bundle"
+	err := os.WriteFile(file1, cnt1, 0600)
 	if err != nil {
-		return fmt.Errorf("unable to parse block number %q: %w", args[0], err)
+		return "", fmt.Errorf("writing temporary file: %w", err)
 	}
+	err = os.WriteFile(file2, cnt2, 0600)
+	if err != nil {
+		return "", fmt.Errorf("writing temporary file: %w", err)
+	}
+
+	cmd := exec.Command("diff", "-u", file1, file2)
+	buffer, _ := cmd.Output()
+
+	out := string(buffer)
+
+	return out, nil
+}
+
+func readBundle(ctx context.Context, filename string, store dstore.Store) ([]string, map[string]*pbeth.Block, error) {
 
-	storeA, err := dstore.NewDBinStore(storeADef)
+	// Create a fileReader
+	fileReader, err := store.OpenObject(ctx, filename)
 	if err != nil {
-		return fmt.Errorf("unable to create store at path %q: %w", storeADef, err)
+		return nil, nil, fmt.Errorf("creating reader: %w", err)
 	}
 
-	storeB, err := dstore.NewDBinStore(storeBDef)
+	// Create a blockReader
+	blockReader, err := bstream.GetBlockReaderFactory.New(fileReader)
 	if err != nil {
-		return fmt.Errorf("unable to create store at path %q: %w", storeBDef, err)
+		return nil, nil, fmt.Errorf("creating block reader: %w", err)
 	}
 
-	blocksA := make(map[string]*pbeth.Block)
-	blocksB := make(map[string]*pbeth.Block)
-
-	filePrefix := fmt.Sprintf("%010d", blockNum)
-	collectBlocks := func(store dstore.Store, blockMap map[string]*pbeth.Block) error {
-		var files []string
-		err = storeA.Walk(ctx, filePrefix, func(filename string) (err error) {
-			files = append(files, filename)
-			return nil
-		})
-
-		var toClose []io.ReadCloser
-		defer func() {
-			for i := range toClose {
-				toClose[i].Close()
-			}
-		}()
+	var blockHashes []string
+	blocksMap := make(map[string]*pbeth.Block)
+	for {
+		curBlock, err := blockReader.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, nil, fmt.Errorf("reading blocks: %w", err)
+		}
 
-		for _, filepath := range files {
-			reader, err := store.OpenObject(ctx, filepath)
-			if err != nil {
-				return err
-			}
-			toClose = append(toClose, reader)
+		curBlockPB := curBlock.ToProtocol().(*pbeth.Block)
+		blockHashes = append(blockHashes, string(curBlockPB.Hash))
+		blocksMap[string(curBlockPB.Hash)] = curBlockPB
+	}
 
-			blockReader, err := bstream.GetBlockReaderFactory.New(reader)
-			if err != nil {
-				return err
-			}
+	return blockHashes, blocksMap, nil
+}
+
+func compareBlocksE(cmd *cobra.Command, args []string) error {
```

Comment:
Generally, you should aim to reduce the size of your method by extracting piece out in simpler functions. To me, `compareBlocksE` feels a bit big and would be good to reduce it a bit. It's a hard thing because it's highly subjective, but here some various element that make me feel it should be done for this method:

- Indentation level is going further than 3-4 levels
- Some chunk of logic seems like they could easily be extracted away, like computing the unified diff between two blocks
- The overall function is a bit long a ~200 lines, it starts to be long IMO going over 150 lines
- Starting around `if displayDiff {`, the logic starts to be harder to reason about which could point to some missing abstractions/functions.

Do **not** do any of those for now, it's more as general comments that I'm saying all this.